### PR TITLE
Create a WeakPtrFactory for use on the UI thread in VsyncWaiter

### PR DESCRIPTION
**Pull request by gw280**
This fixes the failed assertion that the WeakPtr is being used on the wrong thread.

Issue: https://github.com/flutter/flutter/issues/44480

### CODE REVIEW

**Comment by iskakaushik** [comment]
can we write a test for this?

**Comment by gw280** [comment]
It should be covered by the existing Shell unittests, we just don't run them yet on Fuchsia (currently on the roadmap)

**Comment by gw280** [comment]
Oh my bad, Shell unittests use their own implementation of VsyncWaiter for the test harness. I'll have a think about how best to do this but it may end up being put on the tail end of a long list of Fuchsia testing-related items as it'll likely need other work to be completed first (I'm thinking we can probably piggyback off the flutter_runner test harness work).

File: shell/platform/fuchsia/flutter/vsync_waiter.cc [hunk 1]
```diff
@@ -31,13 +34,30 @@ VsyncWaiter::VsyncWaiter(std::string debug_label,
...
+    };
 
   session_wait_.set_handler(wait_handler);
 }
 
 VsyncWaiter::~VsyncWaiter() {
   session_wait_.Cancel();
+
+  fml::AutoResetWaitableEvent ui_latch;
+  fml::MakeCopyable([weak_factory_ui = std::move(weak_factory_ui_),
```
Comment (chinmaygarde) [inline shell/platform/fuchsia/flutter/vsync_waiter.cc]: Don't you need a |RunNowOrPostTask| here for the single threaded configuration?

Comment (gw280) [reply shell/platform/fuchsia/flutter/vsync_waiter.cc]: Oh good catch. I copied the code from the equivalent code in Shell's ctor/dtor. Do we need to change Shell as well?

Comment (chinmaygarde) [reply shell/platform/fuchsia/flutter/vsync_waiter.cc]: I'd think so. Not sure how this wasn't covered by test already. Maybe add an explicit single threaded configuration to the shell_unittests (in a later patch is fine)?

Comment (gw280) [reply shell/platform/fuchsia/flutter/vsync_waiter.cc]: Sorry, I'm not firing on all cylinders today. The Shell code does have it, I just accidentally omitted it.

**Comment by iskakaushik** [comment]
@gw280 given that you have some subset of unittests already working on Fuchsia target device. How feasible is it to add a test to a new test suite that leverages this behavior.

**Comment by gw280** [comment]
> @gw280 given that you have some subset of unittests already working on Fuchsia target device. How feasible is it to add a test to a new test suite that leverages this behavior.

The unittests I'm currently looking at are only exercising the cross-platform code with stub implementations of things like VsyncWaiter so they unfortunately won't cover them.

Luckily though, @dnfield already did some work to add a11y tests at the flutter_runner layer (which this test would need to live in) and it does actually build the Fuchsia VsyncWaiter so it might not be too hard to add the test to that harness. I'll look into that now.

**Comment by gw280** [comment]
@iskakaushik unit test written and seems to work locally!

@chinmaygarde I wasn't thinking straight yesterday, I've updated the changes today and fixed a couple of issues (notably: i was constructing the UI thread's WeakPtrFactory inside a closure instead of the ctor itself, and I was dereffing the WeakPtrFactory on the platform thread in the lambda capture instead of on the UI thread). This has now been resolved and the test passes.

File: shell/platform/fuchsia/flutter/vsync_waiter.cc [hunk 2]
```diff
@@ -57,7 +78,11 @@ void VsyncWaiter::AwaitVSync() {
   fml::TimePoint next_vsync = SnapToNextPhase(now, vsync_info.presentation_time,
                                               vsync_info.presentation_interval);
   task_runners_.GetUITaskRunner()->PostDelayedTask(
-      [self = weak_factory_.GetWeakPtr()] {
+      [& weak_factory_ui = this->weak_factory_ui_] {
+        if (!weak_factory_ui) {
+          return;
```
Comment (iskakaushik) [inline shell/platform/fuchsia/flutter/vsync_waiter.cc]: we should maybe error log here.

File: shell/platform/fuchsia/flutter/vsync_waiter.cc [hunk 3]
```diff
@@ -33,11 +36,29 @@ VsyncWaiter::VsyncWaiter(std::string debug_label,
     FireCallbackNow();
   };
 
+  // Generate a WeakPtrFactory for use with the UI thread. This does not need
+  // to wait on a latch because we only ever use the WeakPtrFactory on the UI
+  // thread so we have ordering guarantees (see ::AwaitVSync())
+  fml::TaskRunner::RunNowOrPostTask(
+      task_runners_.GetUITaskRunner(), fml::MakeCopyable([this]() mutable {
+        this->weak_factory_ui_ =
```
Comment (iskakaushik) [inline shell/platform/fuchsia/flutter/vsync_waiter.cc]: @chinmaygarde should we block this call. I think its ok to not block but jut want to confirm.

Comment (chinmaygarde) [reply shell/platform/fuchsia/flutter/vsync_waiter.cc]: I think this is fine because of the reason mentioned in the comment.

**Pull request merged**